src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs

```diff
@@ -85,133 +122,146 @@ public bool IsSupported(FunctionMetadata functionMetadata, string workerRuntime)
+            var webhostChannels = _languageWorkerChannelManager.GetChannels(_workerRuntime);
+            var workerChannels = webhostChannels == null ? _workerState.GetChannels() : webhostChannels.Union(_workerState.GetChannels());
+            return workerChannels.Where(ch => ch.State == LanguageWorkerChannelState.Initialized);
         }
 
         public void WorkerError(WorkerErrorEvent workerError)
         {
-            if (_workerStates.TryGetValue(workerError.Language, out LanguageWorkerState erroredWorkerState))
+            _logger.LogDebug("Handling WorkerErrorEvent for runtime:{runtime}, workerId:{workerId}", workerError.Language, workerError.WorkerId);
+            _workerState.Errors.Add(workerError.Exception);
+            bool isPreInitializedChannel = _languageWorkerChannelManager.ShutdownChannelIfExists(workerError.Language, workerError.WorkerId);
+            if (!isPreInitializedChannel)
             {
-                _logger.LogDebug("Handling WorkerErrorEvent for runtime:{runtime}", workerError.Language);
-                erroredWorkerState.Errors.Add(workerError.Exception);
-                bool isPreInitializedChannel = _languageWorkerChannelManager.ShutdownChannelIfExists(workerError.Language);
-                if (!isPreInitializedChannel)
-                {
-                    _logger.LogDebug("Disposing errored channel for workerId: {channelId}, for runtime:{language}", erroredWorkerState.Channel.Id, workerError.Language);
-                    erroredWorkerState.Channel.Dispose();
-                }
-                _logger.LogDebug("Restarting worker channel for runtime:{runtime}", workerError.Language);
-                RestartWorkerChannel(workerError.Language, erroredWorkerState);
+                _logger.LogDebug("Disposing errored channel for workerId: {channelId}, for runtime:{language}", workerError.WorkerId, workerError.Language);
+                var erroredChannel = _workerState.GetChannels().Where(ch => ch.Id == workerError.WorkerId).FirstOrDefault();
+                _workerState.DisposeAndRemoveChannel(erroredChannel);
             }
+            _logger.LogDebug("Restarting worker channel for runtime:{runtime}", workerError.Language);
+            RestartWorkerChannel(workerError.Language, workerError.WorkerId);
         }
 
-        private void RestartWorkerChannel(string runtime, LanguageWorkerState erroredWorkerState)
+        private async void RestartWorkerChannel(string runtime, string workerId)
         {
-            if (erroredWorkerState.Errors.Count < 3)
+            if (_workerState.Errors.Count < 3)
             {
-                erroredWorkerState.Channel = CreateNewChannelWithExistingWorkerState(runtime, erroredWorkerState);
-                _workerStates[runtime] = erroredWorkerState;
+                await InitializeJobhostLanguageWorkerChannelAsync(_workerState.Errors.Count);
             }
-            else
+            else if (_workerState.GetChannels().Count() == 0)
```

Comment:
Would it be possible to get into a strange state where we have > 3 exceptions but > 0 workers that got initialized? And what would happen?

Reply:
Yes, this is possible if restarting a failed channel exceeds threshold error count but as long as we have at least one healthy language worker channel, invocations continue to succeed.